vmm/src/vm.rs

```diff
@@ -312,6 +298,27 @@ impl Vm {
         cap.args[0] = ioapic::NUM_IOAPIC_PINS as u64;
         fd.enable_cap(&cap).map_err(Error::VmSetup)?;
 
+        Ok((kvm, fd))
+    }
+
+    pub fn new(
+        config: Arc<Mutex<VmConfig>>,
+        exit_evt: EventFd,
+        reset_evt: EventFd,
+        vmm_path: PathBuf,
+    ) -> Result<Self> {
+        let (kvm, fd) = Vm::kvm_new()?;
+
+        let kernel = File::open(&config.lock().unwrap().kernel.as_ref().unwrap().path)
+            .map_err(Error::KernelFile)?;
+
+        let initramfs = match &config.lock().unwrap().initramfs {
+            Some(initramfs) => Some(File::open(&initramfs.path).map_err(Error::InitramfsFile)?),
+            None => None,
+        };
+
```

Comment:
You can probably do:

```
let initramfs = config.lock().unwrap.initramfs.map(|i| File::open(&i.path).map_err(Error::InitramfsFile)?);
```

Matter of preference really.

Reply:
Just tried but here is the problem:
```
error[E0277]: the `?` operator can only be used in a closure that returns `Result` or `Option` (or another type that implements `std::ops::Try`)                                                                                                                    
   --> vmm/src/vm.rs:319:22
    |
319 |             .map(|i| File::open(&i.path).map_err(Error::InitramfsFile)?);
    |                  ----^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |                  |   |
    |                  |   cannot use the `?` operator in a closure that returns `std::fs::File`
    |                  this function should return `Result` or `Option` to accept `?`
    |
    = help: the trait `std::ops::Try` is not implemented for `std::fs::File`
    = note: required by `std::ops::Try::from_error`

error: aborting due to previous error
```

Reply:
```
let initramfs = config.lock().unwrap.initramfs.map(|i| File::open(&i.path)).transpose().map_err(Error::InitramfsFile)?;
```

This is what `.transpose()` is for.

Reply:
Well you're modifying the behavior if you use transpose, as you'll end up transforming the Result into an Option. The real problem here is that we can't use `?` to return the error from the `File::open()` because it happens inside `map()`.
I don't think we should change the behavior.

Reply:
In my correct version the ? is outside the map. So the .map() returns `Option<Result<T>>` which we then turn into `Result<Option<T>>` with .transpose(). The ? at the end unwraps it and leaves us with just the `Option<T>`

Reply:
Oh yeah I didn't see the trick :)

Reply:
Fixed!